template/conftest.py

```diff
@@ -68,53 +74,66 @@ def pytest_generate_tests(metafunc):
     if "firmware" in metafunc.fixturenames:
         fw_list = []
         ids = []
-        # First pass: enable only demanded firmwares
-        for fw in FIRMWARES:
-            if metafunc.config.getoption(fw.device):
-                fw_list.append(fw)
-                ids.append(fw.device + " " + fw.version)
-        # Second pass if no specific firmware demanded: add them all
-        if not fw_list:
+
+        # If "all" was specified, run on all firmwares
+        if metafunc.config.getoption("all"):
             for fw in FIRMWARES:
                 fw_list.append(fw)
                 ids.append(fw.device + " " + fw.version)
-        metafunc.parametrize("firmware", fw_list, ids=ids)
 
+        else:
+            # Enable only demanded firmwares
+            for fw in FIRMWARES:
+                if metafunc.config.getoption(fw.device):
+                    fw_list.append(fw)
+                    ids.append(fw.device + " " + fw.version)
 
-def prepare_speculos_args(firmware: Firmware, display: bool):
+        if not fw_list:
+            raise ValueError(f"No device specified")
+
+        metafunc.parametrize("firmware", fw_list, ids=ids, scope="session")
+
+
+def prepare_speculos_args(firmware: Firmware, display: bool, elfs_dir: str):
     speculos_args = []
 
     if display:
         speculos_args += ["--display", "qt"]
 
-    app_path = app_path_from_app_name(APPS_DIRECTORY, APP_NAME, firmware.device)
+    if elfs_dir == "bin":
+        app_path = APP_ROOT_DIR / elfs_dir / "app.elf"
+        assert app_path.is_file(), f"{app_path} must exist"
```

Comment:
running a basic `pytest` with no argument in project root will trigger a weird 
`app-X/tests/bin must exist` error?